include/utils/ImplicitSharedDataContainer.h

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2024 UltiMaker
+// CuraEngine is released under the terms of the AGPLv3 or higher
+
+#ifndef UTILS_IMPLICITSHAREDDATACONTAINER_H
+#define UTILS_IMPLICITSHAREDDATACONTAINER_H
+
+namespace cura
+{
+
+template<typename DataType>
+class ImplicitSharedDataContainer
+{
+public:
+    virtual ImplicitSharedDataContainer& operator=(const ImplicitSharedDataContainer& other) noexcept
+    {
+        data_ = other.data_;
+        return *this;
+    }
+
+    virtual ImplicitSharedDataContainer& operator=(ImplicitSharedDataContainer&& other) noexcept
+    {
+        data_ = std::move(other.data_);
+        return *this;
+    }
+
+protected:
+    ImplicitSharedDataContainer() noexcept
+        : data_(std::make_shared<DataType>())
+    {
+    }
+
+    ImplicitSharedDataContainer(const ImplicitSharedDataContainer& other) noexcept = default;
+
+    ImplicitSharedDataContainer(ImplicitSharedDataContainer&& movable) noexcept = default;
+
+    virtual ~ImplicitSharedDataContainer() = default;
+
+    inline const DataType& getData() const noexcept
+    {
+        return *data_;
+    }
+
+    DataType& getData() noexcept
+    {
+        if (! data_.unique())
+        {
+            // Data is shared, make it unique so that we can modify it
+            data_ = std::make_shared<DataType>(*data_);
+        }
+
+        return *data_;
+    }
+
+private:
+    std::shared_ptr<DataType> data_;
+};
+
+template<typename DataType>
+class ImplicitArraydDataContainer : public ImplicitSharedDataContainer<DataType>
+{
+public:
+    inline size_t size() const noexcept
+    {
+        return getData().size();
+    }
+
+    inline bool empty() const noexcept
+    {
+        return getData().empty();
+    }
+
+    DataType::const_reference operator[](size_t index) const
+    {
+        return getData()[index];
+    }
+
+    void reserve(size_t min_size) noexcept
+    {
+        getData().reserve(min_size);
+    }
+
+    /*template<class iterator>
+    ClipperLib::Path::iterator insert(ClipperLib::Path::const_iterator pos, iterator first, iterator last)
+    {
+        return path_->insert(pos, first, last);
+    }*/
```

Comment:
Oh no, that class is a remnant from an architecture I tested... You can safely remove it.

By the way, implicit data sharing is also something we could now easily implement thanks to the encapsulation of the vectors. But from what I could see while profiling, the performance gain would not be huge because most of the time is actually spent in Clipper function. Maybe memory-wise we would have significant gain though.

Reply:
👍🏻